providers/okta/application.go

```diff
@@ -0,0 +1,252 @@
+		}
+	}
+
+	for _, rem := range resp.Remediation.Value {
+		// This is intentionally 'Name' as Type is missing for cases where this is true.
+		if rem.Name == "challenge-authenticator" && resp.CurrentAuthenticatorEnrollment.Value.Key == "duo" {
+			host := resp.CurrentAuthenticatorEnrollment.Value.ContextualData.Host
+			tok, err := ParseSignedToken(resp.CurrentAuthenticatorEnrollment.Value.ContextualData.SignedToken)
+			if err != nil {
+				break
+			}
+
+			return DuoIframe{Host: host, SignedToken: tok, CallbackURL: rem.Href, Method: rem.Method, StateHandle: resp.StateHandle}, true
+		}
+	}
+
+	return nil, false
+}
+
+// ApplicationSAMLSource handles the SP-initiated SAML flow for KeyConjurer.
+type ApplicationSAMLSource string
+
+func (source ApplicationSAMLSource) URL() string {
+	return string(source)
+}
+
+func (source ApplicationSAMLSource) Identify(ctx context.Context, client *http.Client, prevURL *url.URL, username, password string, stateToken StateToken) (identifyResponse IdentifyResponse, err error) {
+	identifyVals := map[string]any{
+		"credentials": map[string]string{
+			"passcode": password,
+		},
+		"identifier":  username,
+		"stateHandle": stateToken,
+	}
+	identifyBytes, _ := json.Marshal(identifyVals)
+	identifyURL := prevURL.ResolveReference(&url.URL{Path: "/idp/idx/identify"})
+	resp, err := client.Post(identifyURL.String(), "application/json", bytes.NewReader(identifyBytes))
+	if err != nil {
+		return
+	}
+
+	if resp.StatusCode == http.StatusForbidden {
+		err = errors.New("user does not have access to this application")
+		return
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(&identifyResponse)
+	return
+}
+
+type IntrospectResponse struct {
+	Success *struct {
+		Name string
+		Href string
+	}
+}
+
+func (source ApplicationSAMLSource) Introspect(ctx context.Context, client *http.Client, prevURL *url.URL, stateToken StateToken) (introspectResponse IntrospectResponse, err error) {
+	// We need to run introspect to get our next link.
+	// Now use /idp/idx/identify to get the IDP link to follow.
+	introspectVals := map[string]any{"stateToken": stateToken}
+	introspectBytes, _ := json.Marshal(introspectVals)
+	introspectURL := prevURL.ResolveReference(&url.URL{Path: "/idp/idx/introspect"})
+	resp, err := client.Post(introspectURL.String(), "application/json", bytes.NewReader(introspectBytes))
+	if err != nil {
+		return
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(&introspectResponse)
+	return
+}
+
+func extractStateToken(r io.Reader) (StateToken, error) {
+	// Extremely cursed
+	bodyBuf, err := io.ReadAll(r)
+	if err != nil {
+		return "", err
+	}
+
+	// Extremely cursed
+	matches := stateTokenExpr.FindSubmatch(bodyBuf)
+	idx := stateTokenExpr.SubexpIndex("Token")
+	if len(matches) < idx+1 {
+		return "", errors.New("no match found")
+	}
+
+	// The JavaScript in the response has HTML-encoded characters in.
+	// When a web browser reads it, this is no problem. But, when we read it, it is.
```

Comment:
I am wondering if there is a Go lib that implements some sort for webview that can parse HTML and run JS code? Probably not, but just curious.

Reply:
What you're describing is something like Selenium WebDriver. This does exist, but the problem with those tools is that they emulate the full browser experience -- including redirects.

KeyConjurer relies on intercepting redirects before they are followed in order to get the SAML response to exchange with AWS. Until we can configure our IdP to let us submit a redirection URL to KeyConjurer when we initiate a saml transaction, we cannot have a browser (headless or otherwise) to do this for us.

This is something that I am investigating for our YubiKey support. With YubiKey, we are forced to use a browser, so we will be figuring out how to pass along the ACS parameter in a SAML request to get us redirected back to, for example, `keyconjurer.example.com/saml2/callback`.